File: 987-vertical-order-traversal-of-a-binary-tree/987-vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        // map ->vertical->level->values
        map<int,map<int,multiset<int>>> nodes; 
        
        queue<pair<TreeNode*,pair<int,int>>> q;
        
        q.push({root,{0,0}});  /// {node,{vertical,level}};
            
        while(!q.empty()){
            
            auto p = q.front();
            q.pop();
            
            TreeNode* currnode = p.first;
            int vertical = p.second.first;
            int level = p.second.second;
            
            nodes[vertical][level].insert(currnode->val);
            
            if(currnode->left){
                q.push({currnode->left,{vertical-1,level+1}});
            }
            
            if(currnode->right){
                q.push({currnode->right,{vertical+1,level+1}});
            }
            
        }
        
        vector<vector<int>> ans;
        
        for(auto p:nodes){
            
            vector<int> col; 
            
            for(auto q:p.second){// iterate over multiset;
                col.insert(col.end(),q.second.begin(),q.second.end()); // insrt at end of col
            }
            ans.push_back(col);
        }
        
        return ans;
    }
};

File: 987-vertical-order-traversal-of-a-binary-tree/987-vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        
        d = defaultdict(list) ## create data structure of
        
        q = [(root,0,0)] # queue
        # node,vertical-level, depth-level
        #node,v,l
        
        while(len(q)!=0):
            
            curr = q[0]
            q.pop(0)
            
            node = curr[0]
            v = curr[1]
            l = curr[2]
            
            if v not in d.keys(): # if v not in d, create dict of dict for v
                d[v] = defaultdict(list)
            
            d[v][l].append(node.val) # add node with main key as v and level as subkey
            
            if(node.left):
                q.append((node.left,v-1,l+1)) # add left right child
            
            if node.right:
                q.append((node.right,v+1,l+1))
        
        # print(d)
                
        ans = []
        
        d= dict(sorted(d.items())) # sort dict acc to main key i.e vertical
        # print(d)

        for v in d.keys(): #iterate over verticl
            
            temp = []
            
            for l in d[v].keys(): # in vertical iterate over level
                d[v][l].sort() # sort the node value in list on same level
                temp.extend(d[v][l])
            
            ans.append(temp)
        
        return ans
            
        
        
        
        
            
            
            
                
            
            
        
        
        